compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/passes/Muxing.kt

```diff
@@ -0,0 +1,401 @@
+import edu.cornell.cs.apl.viaduct.util.FreshNameGenerator
+
+fun ProgramNode.mux(
+    muxedProcesses: Set<Protocol>,
+    muxedFunctions: Set<FunctionName>
+): ProgramNode {
+    val nameAnalysis = NameAnalysis.get(this)
+    return ProgramNode(
+        this.declarations.map { declaration ->
+            when (declaration) {
+                is HostDeclarationNode -> declaration.deepCopy() as TopLevelDeclarationNode
+
+                is ProcessDeclarationNode -> {
+                    if (muxedProcesses.contains(declaration.protocol.value)) {
+                        ProcessDeclarationNode(
+                            declaration.protocol,
+                            declaration.body.mux(nameAnalysis),
+                            declaration.sourceLocation
+                        )
+                    } else {
+                        declaration.deepCopy() as ProcessDeclarationNode
+                    }
+                }
+
+                is FunctionDeclarationNode -> {
+                    if (muxedFunctions.contains(declaration.name.value)) {
+                        FunctionDeclarationNode(
+                            declaration.name,
+                            declaration.pcLabel,
+                            Arguments(
+                                declaration.parameters.map { it.deepCopy() as ParameterNode },
+                                declaration.parameters.sourceLocation
+                            ),
+                            declaration.body.mux(nameAnalysis),
+                            declaration.sourceLocation
+                        )
+                    } else {
+                        declaration.deepCopy() as FunctionDeclarationNode
+                    }
+                }
+            }
+        },
+        this.sourceLocation
+    )
+}
+
+fun ProgramNode.mux(): ProgramNode =
+    this.mux(
+        this.declarations
+            .filterIsInstance<ProcessDeclarationNode>()
+            .map { procDecl -> procDecl.protocol.value }
+            .toSet(),
+        this.functions.map { f -> f.name.value }.toSet()
+    )
+
+fun StatementNode.canMux(): Boolean =
+    when (this) {
+        is LetNode ->
+            when (this.value) {
+                is InputNode -> false
+                is ReceiveNode -> false
+                else -> true
+            }
+
+        is DeclarationNode -> true
+
+        is UpdateNode -> true
+
+        is OutParameterInitializationNode -> false
+
+        is OutputNode -> false
+
+        is SendNode -> false
+
+        is FunctionCallNode -> false
+
+        is IfNode -> this.thenBranch.canMux() && this.elseBranch.canMux()
+
+        is InfiniteLoopNode -> false
+
+        is BreakNode -> false
+
+        is AssertionNode -> false
+
+        is BlockNode -> this.statements.all { child -> child.canMux() }
+    }
+
+fun BlockNode.mux(
+    nameAnalysis: NameAnalysis,
+    nameGenerator: FreshNameGenerator = FreshNameGenerator(nameAnalysis.declaredNames(this))
```

Comment:
This won't avoid parameter names since a `BlockNode` doesn't have access to that information.

Reply:
fixed, the name generator passed when muxing `FunctionDeclarationNode`s is initialized with parameter names